// @vkontakte/vkui/dist/components/ConfigProvider/ConfigProviderContext.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ConfigProviderContext = exports.WebviewType = exports.Scheme = exports.Appearance = void 0;

var _react = require("react");

var Appearance;
exports.Appearance = Appearance;

(function (Appearance) {
  Appearance["DARK"] = "dark";
  Appearance["LIGHT"] = "light";
})(Appearance || (exports.Appearance = Appearance = {}));

var Scheme;
exports.Scheme = Scheme;

(function (Scheme) {
  Scheme["DEPRECATED_CLIENT_LIGHT"] = "client_light";
  Scheme["DEPRECATED_CLIENT_DARK"] = "client_dark";
  Scheme["BRIGHT_LIGHT"] = "bright_light";
  Scheme["SPACE_GRAY"] = "space_gray";
})(Scheme || (exports.Scheme = Scheme = {}));

var WebviewType;
exports.WebviewType = WebviewType;

(function (WebviewType) {
  WebviewType["VKAPPS"] = "vkapps";
  WebviewType["INTERNAL"] = "internal";
})(WebviewType || (exports.WebviewType = WebviewType = {}));

var ConfigProviderContext = (0, _react.createContext)({});
exports.ConfigProviderContext = ConfigProviderContext;
